### ISSUE TITLE
Add an interactive shell debug option for conda build

### ISSUE THREAD

**Issue by yesimon** [conda/conda-build#1751]
Recently I was struggling to debug an erroneous conda build and what I ended up doing was dropping into an interactive shell right before executing `build.sh`. This was really helpful since it allows simulating being in the same environment (environment variables, `$PATH` etc) as the actual build process. This feature exists in the Homebrew package manager as well. 

My suggestion is to add a `--shell` option to drop into an interactive shell right before the building process.

**Comment by msarahan** [comment]
Good idea.  This is a small extension on top of an existing request that I think doesn't have an issue here: recording environment variables in build.sh as is done in bld.bat.

I don't think it'll take the form of dropping into a shell before exiting build.sh - that's just much more complicated.  Instead, what I'd recommend is:

1a) failed builds are already kept.  These live in (miniconda install root)/conda-bld/recipe_folder_name_(timestamp).
1b) for builds that exit successfully, but you still want to keep, use the ``--dirty`` argument.

2) cd into the 'work' folder in that timestamped folder
3) ``source activate ../_b(hit tab to autocomplete this long path)``

basically, a ``--shell`` option should script those actions - but should do so after the build has failed (or succeeded) and exited, and it should pick up where that build left off.  This will just just be simpler from the standpoint of the whole build execution flow.  I think the capability will be equivalent.

**Comment by yesimon** [comment]
In my case the build was succeeding but the binary was incorrect (due to some improper compiler/linker flags). I hope that there's an option to drop in before the build process because sometimes the `build.sh` will change things irrevocably, like deleting/moving source files or changing the environment.

**Comment by msarahan** [comment]
Sorry, I'm just not willing to spend the time to figure out nesting interactive shells in subprocesses.  I'll try to make this kind of feature still available, though, probably with the ``conda develop`` command.  It presently doesn't do very much helpful.  I would like it to create the build env specified by meta.yaml.  It could also set environment variables with activate.d scripts.

If you want to tackle figuring out nesting shells in subprocesses and put up a PR, I'd be glad to check it out.

**Comment by yesimon** [comment]
It seemed to work okay with `subprocess.run("/bin/bash", ...' but I'll definitely investigate a bit deeper to make sure that everything is working properly and getting cleaned up.

**Comment by epruesse** [comment]
At some point, this already worked actually.  It was possible to just put `sh` as a line into build.sh  to drop into a shell exactly at that point (`bash` in `sh` mode will not load any dot-files so leave the environment intact). But it seems that `stdin` is now redirected somewhere. Perhaps just figuring out how/where that happens and circumventing it would suffice to fix this issue.

Alternatively, the solution proposed by @msarahan above would works nicely. Maybe worth adding to the docs somewhere? Or even having a `conda build postmortem` command doing exactly that - opening a new shell inside the last build environment?

**Comment by embray** [comment]
I'm still really struggling to understand how to get into the build environment for a package I'm trying to debug.  See conda-forge/maxima-feedstock#10.  I've tried quite a few things and nothing works, and I feel like if nothing else there should be a documented process for this.
Googling old discussions on the topic has not been helpful since things seem to have changed around in not obvious ways.  For example, I tried running

    conda build --keep-old-work --dirty -m .ci_support/linux_.yaml recipe/

Which kept *something* around but it's not clear what to do with it.  I now have a `/opt/conda/conda-bld/maxima_1538150140061/` directory left behind containing:

```
$ ls
build_prefix_moved_maxima-5.41.0-h0b9cf6d_0_linux-64
pip_cache
prefix_files.txt
_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla
test_tmp
work
work_moved_maxima-5.41.0-h0b9cf6d_0_linux-64
```

where `work/` is empty, but `work_moved.../` does contained the upstream sources, as well as `conda_build.sh`.  In theory (I think?) should be able to source `conda_build.sh` to get into the build environment (this is still annoying because all the environment variables are set along with running the build steps themselves, where really all I need is the environment--perhaps this should be a in separate `source`-able file).

But even that doesn't help, because the remaining `conda_build.sh` contains `export BUILD_PREFIX="/opt/conda/conda-bld/maxima_1538150140061/_build_env"`, which is a path that no longer exists.  There is a `build_prefix_moved_maxima-5.41.0-h0b9cf6d_0_linux-64/`, but that doesn't seem like it either, because just looking around in that directory it seems to just contain dependencies related to the variant I brought in with `-m .ci_support/linux_.yaml`. But it is missing other key build dependencies for this package.  `_build_env/` seems to be gone (despite using `--dirty`, which at least claims to mean "Do not remove work directory or _build environment").

So how exactly do I enable the build environment for a package that's experiencing build and/or test failures?

**Comment by mingwandroid** [comment]
Put exit 1 at the end of your script.

**Comment by msarahan** [comment]
I think you need to take a step back and understand the build process a bit more.

1. recipe is rendered
2. build env (ideally, compilers) created.  _b_env
3. host env (python, libraries, etc) created.  _h_env_placehold...
4. activate build env first, then host env
5. Snapshot files in host env
6. With CWD being work/ run conda_build.sh
7. Snapshot files in host env; bundle new ones
8. Move work folder to the moved name; remove host and build envs (if not --dirty)
9. create test env
10. run package's test scripts
11. If tests fail, move the package to the "broken" subdir
12. Try to clean up build folder (pkgname_timestamp)


> So how exactly do I enable the build environment for a package that's experiencing build and/or test failures?

As Ray mentions, ``exit 1`` at the end of your script is often a good way to get the build to stop at the state that you want it.  When builds fail, you'll still have the "work" dir and the build and host envs.  You then need to activate the build and host envs.  Take a look in conda_build.sh - it has code there that does this.  It also has code to set env vars.

As I write this, I really think we could do a lot by just separating the activation and env var setting into another script that gets sourced by the conda-build script.  With that activation/env var script, you could just source it yourself and be ready for debugging.

Tests are much the same, but you only have the _t_env to activate.  The test_tmp folder is the CWD for tests.

**Comment by embray** [comment]
Thank you for the details, but I'm still confused about some things.  For example:

> Move work folder to the moved name; remove host and build envs (if not --dirty)

But I did pass `--dirty` and it still seems to remove those environments.

I'm also not sure about the `exit 1` stuff and forcing a "failed" build.  Just in principle I think manually forcing the build to "fail" shouldn't be necessary to do some post-mortem debugging (as already noted elsewhere on this issue).  There is a `--keep-old-work` which the docs state "Do not remove anything from environment, even  after  successful build and test."   But it's not clear what it's "not removing".  It didn't seem to do anything useful to me.

> As I write this, I really think we could do a lot by just separating the activation and env var setting into another script that gets sourced by the conda-build script. With that activation/env var script, you could just source it yourself and be ready for debugging.

Yes, that was exactly what I was suggesting when I wrote (parenthetically) "perhaps this should be a in separate source-able file" :)  Even then, that won't work if the `_build_env` gets removed, and it's very confusing that that seems to be happening with any combination of flags I pass to `conda build`.

Also what is `build_prefix_moved...`?

**Comment by embray** [comment]
> We move build-time folders out of the way before testing deliberately so that any hard-coded references to files within them will fail, indicating we need to somehow fix that.

That makes sense in general.  But it would be nice to have an option (again, as already suggested in this issue) for development purposes to be able to be able to interrupt that process elsewhere.  Most other packaging systems I've worked with have something like this, I feel like.

> We're not able to spend time on this, and I also find exit 1 (and have done for more than a year) to be entirely sufficient, and appropriate for my needs.

I'd be happy to supply a PR but I think I'll need to understand more first.    That may be that it's entirely sufficient, but it's certainly not *obvious*, and it's not documented anywhere how to debug and develop packages.  Or at the very least, it would be nice if there were a way I could find to just create a build+host environment from a package's meta.yaml and then manually run `build.sh` in that environment.  If there's a way to do that it's not obvious either.  It's also not clear from the docs what `--dirty` or `--keep-old-work` actually *do*, but in practice neither are what I'm looking for it seems.

**Comment by mingwandroid** [comment]
The thing is, at whatever point you want to interrupt the build process, you need to do something to that line in build.sh (or wherever), so whatever mechanism we could come up with (some shell function like breakpoint()) is going to be no less effort than inserting `exit 1`.

We could document this, again PR welcome, I personally feel like it would be documenting a generic unix-ism though.

**Comment by ihnorton** [comment]
@embray [this project](https://github.com/ESSS/conda-devenv) might be of interest. I haven't used it yet [1], but it seems to be geared toward what you want.

[1] my own debugging process so far is: activate the _placeholder env, then edit a copy of the `conda_build.sh` script to leave only exports, and source it

**Comment by embray** [comment]
> The thing is, at whatever point you want to interrupt the build process, you need to do something to that line in build.sh (or wherever), so whatever mechanism we could come up with (some shell function like breakpoint or something) is going to be no less effort (for the end user, but considerable effort for whoever implements this) than inserting exit 1.

I feel like we're talking about two different things here.  You had mentioned that the build has separate "build" and "test" phases.  Of course, a `build.sh` can have something in it like `make check`.  But how is that distinct from the "test" phase you're describing when you write "Tests are much the same, but you only have the _t_env to activate. The test_tmp folder is the CWD for tests."

What I'm saying is, if I'm just trying to debug a package--maybe I want to manually re-run `make` or `make check`, etc--I don't care about above moving things around in support of this.  I just want to drop into the build environment as it was at the end of a successful or unsuccessful `build.sh` run.

Obviously if I wanted to interrupt `build.sh` somewhere specific I could do that.  What is the point of `--dirty` and/or `--keep-old-work` if they don't keep the build environment intact after a successful build?  It's not at all obvious that inserting a random `exit 1` into the script would be the solution to the use case I'm trying to describe.

> We could document this, again PR welcome, I personally feel like it would be documenting a generic unix-ism though.

I meant documenting both what @msarahan wrote above, as well as in general the technique for debugging a build in conda which is clearly *not* obvious if I, as someone with experience with other packaging systems and having read all the docs for conda and conda-forge, cannot figure out how to do something which should be relatively simple.

I'm not levying a criticism here; I have plenty of my own projects where things that should be simple are not obvious or well-documented.  I just think it's a valid problem, and that maybe you're misunderstanding exactly what my problem is...

**Comment by embray** [comment]
Also, it was suggested to me by @isuruf, as well as @epruesse on this issue, that I can just put `bash` or `sh` somewhere in the build script and drop in that way.  It's still non-obvious that this should work at all, but it seems like a simple enough solution to my problem.  However, when I tried that, `conda build` just hanged.  I'm not sure exactly why, though I guess maybe the `build.sh` was run in such a manner that `bash` couldn't find a TTY to attach to and is just stuck waiting for stdin or something.  It wasn't immediately clear though.  Ctrl-C crashed `conda builda` with a `KeyboardInterrupt` at that point.

**Comment by mingwandroid** [comment]
The sh method has problems to do with nested shells as @msarahan pointed out.

**Comment by embray** [comment]
I think what would go a long way would be to help me understand *exactly* what the flags `--dirty` and `--keep-old-work` do, or are supposed to do anyways.

I just discovered that `--dirty` also sets the `DIRTY=1` environment variable which can be used in build recipes, but it's not obvious that that's all it does.

**Comment by mingwandroid** [comment]
It would be awesome if we could hook up the [bash debugger](http://bashdb.sourceforge.net/) somehow.

**Comment by embray** [comment]
Whoa, I've never used that before but it looks highly useful.  I think it's still above and beyond what I need or want right now w.r.t. conda but in general that would be a handy thing to have.  I'm less worried about debugging the `build.sh` scripts themselves, and more with investigating bugs and test failures that are only appearing in conda builds (again, I got into this issue because I wanted to investigate conda-forge/maxima-feedstock#10, though in the past I have *tried* to help out more with packaging for conda-forge only to be stymied by this same problem...)

**Comment by mingwandroid** [comment]
> As suggested, I edited the recipe's build.sh to include exit 1 (this assumes the recipe even has a build.sh which is not necessary for many packages)

Even when there's no recipe you can use multiple `build/script` commands so this is not a problem.

> I copied conda_build.sh and renamed it conda_env.sh. I edited conda_build.sh back down to the original build.sh without any of the environment setup (or the exit 1), whereas I edited conda_env.sh to just contain the environment setup.

Changes to conda-build so that we have conda_env.sh natively created and sourced from conda_build.sh would be appreciated, but it should include the activation of the environments.

**Comment by embray** [comment]
> If you run conda deactivate beforehand you don't need to mess about with this stuff, I suspect this is a conda bug though and should be investigated.

That makes some sense, I think.

> but it should include the activation of the environments in conda_env.sh

Yes, I think that goes without saying, unless I'm misinterpreting.  To me the whole point of this exercise was to activate said environments so that I could play around in them.  Tbh a pretty workable alternative as well might be a simple way to create and activate the host+build environments specified by a meta.yaml just as a normal conda environment.  I think that's the majority of what I'm trying to get at here...

**Comment by mingwandroid** [comment]
> Tbh a pretty workable alternative as well might be a simple way to create and activate the host+build environments specified by a meta.yaml just as a normal conda environment.

We'll that env wouldn't contain your source, built and intermediate files. I think we should try to make things more ergonomic in small steps and splitting conda_build.sh would be a good first step.

**Comment by msarahan** [comment]
WIP implementation at https://github.com/conda/conda-build/pull/3237

Feedback welcome.
